src/storage/mvcc/txn.rs

```diff
@@ -728,32 +752,40 @@ impl<S: Snapshot> MvccTxn<S> {
                 }
             }
             None => {
-                if rollback_if_not_exist {
-                    let ts = self.start_ts;
-
-                    // collapse previous rollback if exist.
-                    if self.collapse_rollback {
-                        self.collapse_prev_rollback(primary_key.clone())?;
-                    }
-
-                    // Insert a Rollback to Write CF in case that a stale prewrite
-                    // command is received after a cleanup command.
-                    let write = Write::new_rollback(ts, protect_rollback);
-                    self.put_write(primary_key, ts, write.as_ref().to_bytes());
-                    MVCC_CHECK_TXN_STATUS_COUNTER_VEC.rollback.inc();
-
-                    Ok(TxnStatus::LockNotExist)
-                } else {
-                    Err(ErrorInner::TxnNotFound {
+                if MissingLockAction::ReturnError == action {
+                    return Err(ErrorInner::TxnNotFound {
                         start_ts: self.start_ts,
                         key: primary_key.into_raw()?,
                     }
-                    .into())
+                    .into());
+                }
+
+                let ts = self.start_ts;
+
+                // collapse previous rollback if exist.
+                if self.collapse_rollback {
+                    self.collapse_prev_rollback(primary_key.clone())?;
                 }
+
+                // Insert a Rollback to Write CF in case that a stale prewrite
+                // command is received after a cleanup command.
+                let write = self.get_write(action, ts);
+                self.put_write(primary_key, ts, write.as_ref().to_bytes());
+                MVCC_CHECK_TXN_STATUS_COUNTER_VEC.rollback.inc();
+
+                Ok(TxnStatus::LockNotExist)
             }
         }
     }
 
+    fn get_write(&self, action: MissingLockAction, ts: TimeStamp) -> Write {
```

Comment:
It's not a proper name and I can't think of one. How about adding a method to `MissingLockAction`:
```rust
impl MissingLockAction {
...
    fn into_write/or rollback(self, ts: TimeStamp) -> Write {
        match self {
            MissingLockAction::Rollback => Write::new_rollback(ts, false),
            MissingLockAction::ProtectedRollback => Write::new_rollback(ts, true),
            _ => unreachable!(),
        }
    }
}